common/src/seg_tree.c

```diff
@@ -189,29 +213,43 @@ int seg_tree_add(struct seg_tree* seg_tree, unsigned long start,
             /* Insert the non-overlapping part of the new range */
             RB_INSERT(inttree, &seg_tree->head, resized);
 
+            /* if the non-overlapping part came from the front
+             * portion of the existing range, then there is a
+             * trailing portion of the existing range to add back
+             * to be considered again in the next loop iteration */
             if (resized->end + 1 >= overlap->start &&
-                resized->end +1 <= overlap->end) {
+                resized->end + 1 <= overlap->end) {
                 /*
                  * There's still a remaining section after the non-overlapping
                  * part.  Add it in.
                  */
-                remaining = seg_tree_node_alloc(resized->end + 1, overlap->end,
+                struct seg_tree_node* remaining = seg_tree_node_alloc(
+                    resized->end + 1, overlap->end,
                     overlap->ptr + (resized->end + 1 - overlap->start));
-                if (!resized) {
+                if (!remaining) {
                     free(overlap);
                     return ENOMEM;
                 }
                 RB_INSERT(inttree, &seg_tree->head, remaining);
                 seg_tree->count++;
             }
+
+            /* we've extracted this node from the tree and inserted
+             * newly allocated nodes to replace it, free the old node */
             free(overlap);
         }
     }
+
+    /* increment segment count in the tree for the range we just added */
     if (!overlap) {
         seg_tree->count++;
     }
```

Comment:
I think (!overlap) should always be true for us to have exited the while loop, right?  If so, we can drop the if() portion and just increment the count.